samples/langchain_quick_start.ipynb

```diff
@@ -20,14 +20,340 @@
+    "\n",
+    "chat_history = MemorystoreChatMessageHistory(\n",
+    "    client=client,\n",
+    "    session_id=\"my_session\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Build a Movie Question-Answering Chatbot"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_google_vertexai import VertexAIEmbeddings, VertexAI\n",
+    "from langchain_core.messages import AIMessage, HumanMessage\n",
+    "from langchain.chains import ConversationalRetrievalChain\n",
+    "from langchain.memory import ConversationSummaryBufferMemory\n",
+    "from langchain_core.prompts import PromptTemplate\n",
+    "from langchain_google_memorystore_redis import MemorystoreChatMessageHistory\n",
+    "\n",
+    "# Suppress all deprecation warnings\n",
+    "import warnings\n",
+    "\n",
+    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
+    "\n",
+    "# Prepare some prompt templates for the ConversationalRetrievalChain\n",
+    "prompt = PromptTemplate(\n",
+    "    template=\"\"\"Use all the information from the context and the conversation history to answer new question. If you see the answer in previous conversation history or the context. \\\n",
+    "Answer it with clarifying the source information. If you don't see it in the context or the chat history, just say you \\\n",
+    "didn't find the answer in the given data. Don't make things up.\n",
+    "\n",
+    "Previous conversation history from the questioner. \"Human\" was the user who's asking the new question. \"Assistant\" was you as the assistant:\n",
+    "```{chat_history}\n",
+    "```\n",
+    "\n",
+    "Vector search result of the new question:\n",
+    "```{context}\n",
+    "```\n",
+    "\n",
+    "New Question:\n",
+    "```{question}```\n",
+    "\n",
+    "Answer:\"\"\",\n",
+    "    input_variables=[\"context\", \"question\", \"chat_history\"],\n",
+    ")\n",
+    "condense_question_prompt_passthrough = PromptTemplate(\n",
+    "    template=\"\"\"Repeat the following question:\n",
+    "{question}\n",
+    "\"\"\",\n",
+    "    input_variables=[\"question\"],\n",
+    ")\n",
+    "\n",
+    "# Intialize retriever, llm and memory for the chain\n",
+    "retriever = vector_store.as_retriever(\n",
+    "    search_type=\"mmr\", search_kwargs={\"k\": 5, \"lambda_mult\": 0.8}\n",
+    ")\n",
+    "llm = VertexAI(model_name=\"gemini-pro\", project=f\"{PROJECT_ID}\")\n",
+    "\n",
+    "chat_history.clear()\n",
+    "\n",
+    "memory = ConversationSummaryBufferMemory(\n",
+    "    llm=llm,\n",
+    "    chat_memory=chat_history,\n",
+    "    output_key=\"answer\",\n",
+    "    memory_key=\"chat_history\",\n",
+    "    return_messages=True,\n",
+    ")\n",
+    "\n",
+    "# create the ConversationalRetrievalChain\n",
+    "rag_chain = ConversationalRetrievalChain.from_llm(\n",
```

Comment:
right before # create the conversations retreival chain add in text block
"Now let's create a conversational retrieval chain. This will allow the LLM to use chat history in it's responses, meaning we can ask it follow up questions to our questions instead of having to start from scratch after each inquiry."

Reply:
this code snippet has been broken up into smaller chunks and the writing has been updated to match the guidelines. ptal